[GTK][WPE] Don't request more than a single composition during a layerFlush

magomez opened the pull request:
#### 1dcc05917fd767be73b5b47c69036eae2a76fba9
<pre>
[GTK][WPE] Don&apos;t request more than a single composition during a layerFlush
<a href="https://bugs.webkit.org/show_bug.cgi?id=272324">https://bugs.webkit.org/show_bug.cgi?id=272324</a>

Reviewed by Carlos Garcia Campos.

If there weren&apos;t changes to the page that requested a composition with a new state,
request a composition of the current content.

* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::layerFlushTimerFired):

Canonical link: <a href="https://commits.webkit.org/277234@main">https://commits.webkit.org/277234@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/beba271050fcfbc8cda66176bf377e477ec27820

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46969 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26136 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49651 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43017 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31249 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23603 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38258 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47550 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40469 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19567 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20930 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5015 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43349 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42003 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51525 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21988 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18445 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45553 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23271 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44539 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10388 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24040 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22982 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/797cf8e830617695d76fdef9c2ffe82d5cb917b6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26970

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
```diff
@@ -179,7 +179,9 @@ void LayerTreeHost::layerFlushTimerFired()
 
 #if HAVE(DISPLAY_LINK)
-    m_compositor->updateScene();
+    // Request a composition only if the CompositingCoordinator hasn't called commitSceneState.
+    if (!m_isWaitingForRenderer)
+        m_compositor->updateScene();
```
This looks good to me, but to understand it you need to check that `m_isWaitingForRenderer` is set to true in `LayerTreeHost::commitSceneState` that is called from `CompositingCoordinator::flushPendingLayerChanges` only when `shouldSyncFrame` is true. So, I wonder if we can make this easier to follow. I see two options:

* `CompositingCoordinator::flushPendingLayerChanges` returns a boolean that is never read and always returns true. We could return `shouldSyncFrame` and use that value to decide in `LayerTreeHost::layerFlushTimerFired()`
* We could remove `m_compositor->updateScene()` from `LayerTreeHost::layerFlushTimerFired()` and call `m_client.updateScene()` from `CompositingCoordinator::flushPendingLayerChanges` when `shouldSyncFrame` is false

magomez replied:
> Maybe second option makes more sense, I wonder if it's important to call `updateScene` before `m_page.didUpdateRendering()`, since `commitSceneState()` is called before.

Yes, that makes it easier to understand.
I think `updateScene` can be called before or after `m_page.didUpdateRendering()` without much trouble.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/beba271050fcfbc8cda66176bf377e477ec27820. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/26970

magomez committed:
[GTK][WPE] Don't request more than a single composition during a layerFlush
https://bugs.webkit.org/show_bug.cgi?id=272324

Reviewed by Carlos Garcia Campos.

If there weren't changes to the page that requested a composition with a new state,
request a composition of the current content.

* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::layerFlushTimerFired):

Canonical link: https://commits.webkit.org/277234@main

webkit-commit-queue commented:
Committed 277234@main (1dcc05917fd7): <https://commits.webkit.org/277234@main>

Reviewed commits have been landed. Closing PR #26970 and removing active labels.
